operations/secure_random/secure_random.cc

```diff
@@ -0,0 +1,203 @@
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/kernels/bounds_check.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/lib/random/random_distributions.h"
+#include "sodium.h"
+
+using namespace tensorflow;
+
+using shape_inference::DimensionHandle;
+using shape_inference::ShapeHandle;
+
+#define CHACHABLOCKSIZE 64
+
+static Status SecureRandomShape(shape_inference::InferenceContext* context) {
+  // Check seed shape
+  ShapeHandle seed;
+  TF_RETURN_IF_ERROR(context->WithRank(context->input(1), 1, &seed));
+  DimensionHandle unused;
+  TF_RETURN_IF_ERROR(context->WithValue(context->Dim(seed, 0), 8, &unused));
+
+  // Set output shape
+  ShapeHandle out;
+  TF_RETURN_IF_ERROR(context->MakeShapeFromShapeTensor(0, &out));
+  context->set_output(0, out);
+  return Status::OK();
+}
+
+REGISTER_OP("SecureRandom")
+    .Input("shape: T")
+    .Input("seed: Tseed")
+    .Input("minval: dtype")
+    .Input("maxval: dtype")
+    .Output("output: dtype")
+    .Attr("dtype: {int32, int64} = DT_INT32")
+    .Attr("T: {int32, int64} = DT_INT32")
+    .Attr("Tseed: {int32} = DT_INT32")
+    .SetShapeFn(SecureRandomShape);
+
+// this function allows you to skip ahead in the chacha stream so you don't
+// have to allocate more memory than you need to, used in rejection sampling and
+// will be handy for parallelizing this operation
+void randombytes_buf_deterministic_ic(void * const buf, const size_t size, uint32_t count,
+                              const unsigned char seed[randombytes_SEEDBYTES])
+{
+    static const unsigned char nonce[crypto_stream_chacha20_ietf_NONCEBYTES] = {
+        'L', 'i', 'b', 's', 'o', 'd', 'i', 'u', 'm', 'D', 'R', 'G'
+    };
+
+    unsigned char * u_buf = (unsigned char *)buf;
+
+    memset(u_buf, 0, size);
+
+    crypto_stream_chacha20_ietf_xor_ic(u_buf, u_buf, (unsigned long long) size,
+                                       nonce, count, seed);
+}
+
+
+template <typename T>
+class Generator {
+public:
+  Tensor *output = NULL;
+  const unsigned char * seeds = NULL;
+
+  Generator(Tensor* output, const unsigned char * seeds) : output(output), seeds(seeds) {
+    auto flat = output->flat<T>();
+
+    count_ = output->flat<T>().size();
+    bytes_count_ = count_ * sizeof(T);
+    buf_ = static_cast<T *>(malloc(bytes_count_));
```

Comment:
is there anyway we can avoid this `malloc` and use what was already allocated by TF for the output (which might keep a memory pool internally)? or perhaps it doesn't work that way..?

Reply:
Good idea. Thing I'm stuck with here is that the output data from `output->flat<T>()` returns a `std::vector` I believe. I couldn't figure out a good way to access the underlying memory allocated for that. We need to do that because libsodium takes a void pointer. We can't just pass a vector through. Any ideas?

Reply:
my C++ is *very* limited, best I can do is stackoverflow :D https://stackoverflow.com/questions/6485496/how-to-get-stdvector-pointer-to-the-raw-data

Reply:
This appears to work! Great find!